cli/command/manifest/push.go

```diff
@@ -0,0 +1,591 @@
+		return []string{}, nil
+	}
+
+	if jsonData != nil {
+		cf := configfile.ConfigFile{}
+		if err := json.Unmarshal(jsonData, &cf); err != nil {
+			logrus.Debugf("manifest create: Unable to unmarshal insecure registries from $HOME/.docker/config.json: %s", err)
+			return []string{}, nil
+		}
+		if cf.InsecureRegistries == nil {
+			return []string{}, nil
+		}
+		// @TODO: Add tests for a) specifying in config.json, b) invalid entries
+		for _, reg := range cf.InsecureRegistries {
+			if err := net.ParseIP(reg); err == nil {
+				insecureRegistries = append(insecureRegistries, reg)
+			} else if _, _, err := net.ParseCIDR(reg); err == nil {
+				insecureRegistries = append(insecureRegistries, reg)
+			} else if ips, err := net.LookupHost(reg); err == nil {
+				insecureRegistries = append(insecureRegistries, ips...)
+			} else {
+				return []string{}, errors.Wrapf(err, "manifest create: Invalid registry (%s) specified in ~/.docker/config.json: %s", reg)
+			}
+		}
+	}
+
+	return insecureRegistries, nil
+}
+
+func pushReferences(httpClient *http.Client, urlBuilder *v2.URLBuilder, ref reference.Named, manifests []manifestPush) error {
+	for _, manifest := range manifests {
+		dgst, err := digest.Parse(manifest.Digest)
+		logrus.Debugf("pushing ref digest %s", dgst)
+		if err != nil {
+			return errors.Wrapf(err, "error parsing manifest digest (%s) for referenced manifest %q: %v", manifest.Digest, manifest.Name)
+		}
+		targetRef, err := reference.WithDigest(ref, dgst)
+		logrus.Debugf("pushing ref %v", targetRef)
+		if err != nil {
+			return errors.Wrapf(err, "error creating manifest digest target for referenced manifest %q: %v", manifest.Name)
+		}
+		pushURL, err := urlBuilder.BuildManifestURL(targetRef)
+		if err != nil {
+			return errors.Wrapf(err, "error setting up manifest push URL for manifest references for %q: %v", manifest.Name)
+		}
+		logrus.Debugf("manifest reference push URL: %s", pushURL)
+
+		pushRequest, err := http.NewRequest("PUT", pushURL, bytes.NewReader(manifest.JSONBytes))
+		if err != nil {
+			return errors.Wrap(err, "HTTP PUT request creation for manifest reference push failed")
+		}
+		pushRequest.Header.Set("Content-Type", manifest.MediaType)
+		resp, err := httpClient.Do(pushRequest)
+		if err != nil {
+			return errors.Wrap(err, "PUT of manifest reference failed")
+		}
+
+		resp.Body.Close()
+		if !statusSuccess(resp.StatusCode) {
+			return fmt.Errorf("referenced manifest push unsuccessful: response %d: %s", resp.StatusCode, resp.Status)
+		}
+		dgstHeader := resp.Header.Get("Docker-Content-Digest")
+		dgstResult, err := digest.Parse(dgstHeader)
+		if err != nil {
+			return errors.Wrap(err, "couldn't parse pushed manifest digest response")
+		}
+		if string(dgstResult) != manifest.Digest {
+			return fmt.Errorf("pushed referenced manifest received a different digest: expected %s, got %s", manifest.Digest, string(dgst))
```

Comment:
I don't think this should be treated as an error. For example, the registry could use a different digest algorithm. Currently everything uses sha256, but I believe the API is flexible.